Implement PEP 788

ZeroIntensity opened the pull request:
Reference implementation for PEP-788.

ZeroIntensity committed:
Implement interpreter state reference counting.

ZeroIntensity committed:
Add a test for interpreter reference counts.

ZeroIntensity committed:
Add thread state daemon-ness that doesn't work yet.

ZeroIntensity committed:
Add untested implementation of non-daemon native threads.

ZeroIntensity committed:
Add a test for PyThreadState_SetDaemon().

ZeroIntensity committed:
Add untested implementation of Ensure()/Release() that probably doesn't work and isn't thread-safe.

ZeroIntensity committed:
Change some comments.

ZeroIntensity committed:
Add a test that I'm sure doesn't work.

ZeroIntensity committed:
Use the interpreter's reference count and native thread countdown as one.

ZeroIntensity committed:
Fix the countdown decrement.

ZeroIntensity committed:
Remove unused variable.

ZeroIntensity committed:
Test for PyGILState_Ensure()

ZeroIntensity committed:
Fix the test for the new reference counting.

ZeroIntensity committed:
Add PyInterpreterState_Lookup()

ZeroIntensity committed:
Fix a few bugs and add a test.

ZeroIntensity committed:
Add a test for PyThreadState_Ensure() across interpreters.

ZeroIntensity committed:
Remove an artifact from old approach.

colesbury commented on Python/pylifecycle.c:
```diff
@@ -2428,6 +2433,9 @@ Py_EndInterpreter(PyThreadState *tstate)
     wait_for_thread_shutdown(tstate);
 
+    // Wrap up non-daemon native threads
+    wait_for_native_shutdown(tstate->interp);
```
I think this needs to be unified with `wait_for_thread_shutdown`. Threads created from Python may spawn threads in C and vice versa. 

As currently written, I think you can get to `wait_for_native_shutdown()` and then have a thread spawn a `threading.Thread()` that's not waited on.

ZeroIntensity replied:
Yeah, I thought this might be an issue. Especially since there seems to be some inclination for a `PyThreadState_GetDaemon` function, unifying these seems like a good idea.

My main concern is breakage towards people who are manually using `threading._shutdown` for whatever reason. We'd have to remove that if we treat `threading` threads as native threads (or I guess we could have `threading._shutdown` call `wait_for_native_shutdown`?).

ZeroIntensity replied:
I think this problem is out of the scope of PEP 788. Pending calls or atexit handlers can also start threads, so we need to fix this in some other way. I think a loop that runs all these finalizer functions until there's nothing called anymore should work.

colesbury replied:
I disagree. Spawning threads from other threads is a very common operation. Spawning threads from atexit handlers is not.

ZeroIntensity replied:
Right, but we should fix it anyway. You can get some nifty fatal errors if you do something like this right now:

```py
import atexit
import threading

@atexit.register
def start_thread():
    threading.Thread(target=print, args=("hello world",)).start()
```

My point is that we should implement that fix regardless of PEP 788, and then update the reference implementation once it lands. Is that ok?

ZeroIntensity replied:
FYI, this is now tracked at gh-136003.

ZeroIntensity committed:
Fix test from earlier semantics.

colesbury commented on Python/pystate.c:
```diff
@@ -3185,3 +3218,165 @@ _Py_GetMainConfig(void)
+        return -1;
+    }
+    PyThreadState *save = _PyThreadState_GET();
+    if (save != NULL && save->interp == interp) {
```
One of the properties of `PyGILState_Ensure()` was that if you hade code that did something like:

```
Py_BEGIN_ALLOW_THREADS
...
PyGILState_Ensure();
```

Then the `PyGILState_Ensure()` code would reuse the same thread state; it wouldn't create a new one just because the thread state isn't active. The equivalent code with `PyThreadState_Ensure()` would create a new thread state.

This can come up when you release the GIL and make a possibly long running call into a C library function, which then calls back into Python.

* I didn't see this in the PEP. I think it's worth mentioning if it's not already there.
* Creating thread states unnecessarily can be expensive
* Creating a new thread state means that you have different thread local variables and other weird behavior.

ZeroIntensity replied:
Ok, we can reuse the gilstate pointer for this case. That has the additional benefit of being compatible with `PyGILState_Ensure` at the same time. Thanks for bringing this up.

(I'll update the PEP in a big "round 1 comments" PR sometime tomorrow or in the next few days. I want to give people a chance to look at the current draft first, so I can get a good idea of what needs to change.)

colesbury commented on Python/pystate.c:
```diff
@@ -1806,6 +1815,26 @@ PyThreadState_Clear(PyThreadState *tstate)
+}
+
+static void
+decrement_daemon_count(PyInterpreterState *interp)
```
minor, but this decrements the count of *non-daemon* threads, right? It's not a count of *daemon* threads.

ZeroIntensity replied:
Oh yeah, oops.

vstinner commented on Python/pylifecycle.c:
```diff
@@ -3454,6 +3462,25 @@ wait_for_thread_shutdown(PyThreadState *tstate)
+    }
+    PyMutex_Unlock(&finalizing->mutex);
+
+    PyEvent_Wait(&finalizing->finished);
```
Is it possible to interrupt it with CTRL+C? If not, maybe PyEvent_WaitTimed() should be called in a loop and call PyErr_CheckSignals() time to time?

vstinner commented on Python/pystate.c:
```diff
@@ -3185,3 +3218,165 @@ _Py_GetMainConfig(void)
+PyThreadState_Ensure(PyInterpreterState *interp)
+{
+    assert(interp != NULL);
+    _Py_ensured_tstate *entry = PyMem_RawMalloc(sizeof(_Py_ensured_tstate));
```
You might rename 'entry' to 'ensured'.

ZeroIntensity committed:
Remove 'daemonness' as a property of a thread.

ZeroIntensity committed:
Add strong interpreter reference functions.

ZeroIntensity committed:
Implement weak references.

ZeroIntensity committed:
Fix some thread safety issues regarding interpreter deletion.

ZeroIntensity committed:
Merge from main branch.

ZeroIntensity committed:
Implement new version of PyThreadState_Ensure() and PyThreadState_Release()

ZeroIntensity committed:
Use the new APIs in the tests.

ZeroIntensity committed:
Fix _testcapi.

ZeroIntensity committed:
Fix the ensure counter.

ZeroIntensity committed:
Add the test to test_embed.

ZeroIntensity committed:
Allow the wait to be interrupted by CTRL+C.

ZeroIntensity committed:
Print the error before bailing out.

ZeroIntensity committed:
Updates for the new proposal.

ZeroIntensity committed:
Bikeshedding.

vstinner commented on Include/cpython/pystate.h:
```diff
@@ -259,3 +265,45 @@ PyAPI_FUNC(_PyFrameEvalFunction) _PyInterpreterState_GetEvalFrameFunc(
+#define PyInterpreterRef_Close(ref) do {    \
+    PyInterpreterRef_Close(ref);            \
+    ref = 0;                                \
+} while (0);
```
```suggestion
} while (0)
```

vstinner commented on Include/cpython/pystate.h:
```diff
@@ -259,3 +265,45 @@ PyAPI_FUNC(_PyFrameEvalFunction) _PyInterpreterState_GetEvalFrameFunc(
+
+/* Weak interpreter references */
+
+typedef struct _interpreter_weakref {
```
```suggestion
typedef struct _PyInterpreterWeakRef {
```

vstinner commented on Include/cpython/pystate.h:
```diff
@@ -259,3 +265,45 @@ PyAPI_FUNC(_PyFrameEvalFunction) _PyInterpreterState_GetEvalFrameFunc(
+#define PyInterpreterWeakRef_Close(ref) do {    \
+    PyInterpreterWeakRef_Close(ref);            \
+    ref = 0;                                    \
+} while (0);
```
```suggestion
} while (0)
```

vstinner commented on Include/cpython/pystate.h:
```diff
@@ -259,3 +265,45 @@ PyAPI_FUNC(_PyFrameEvalFunction) _PyInterpreterState_GetEvalFrameFunc(
+} while (0);
+
+// Exports for '_testcapi' shared extension
+PyAPI_FUNC(Py_ssize_t) _PyInterpreterState_Refcount(PyInterpreterState *interp);
```
Can you move it to the internal C API? (Include/internal/pycore_pystate.h).

Same remark for _PyInterpreterState_Incref().

ZeroIntensity replied:
`_testcapi` doesn't allow importing from pycore headers. I only need these for those tests.

vstinner replied:
Move tests to _testinternalcapi in this case.

ZeroIntensity replied:
But we're not testing the internal C API, we're testing the public C API. We just use the internal C API to write those tests.

vstinner replied:
Please don't add private functions (ex: `_PyInterpreterState_Refcount()`) to the public C API. So move it to the internal C API. Sadly, it means moving your tests to the internal C API tests, unless a test only uses public functions.

ZeroIntensity replied:
I split up the tests, so now the internal stuff is in `_testinternalcapi` and the internal functions are in the private headers.

vstinner commented on Lib/test/test_embed.py:
```diff
@@ -1919,6 +1919,10 @@ def test_gilstate_after_finalization(self):
         self.run_embedded_interpreter("test_gilstate_after_finalization")
 
 
```
We use a single empty line between test case methods. Please remove one empty line.

ZeroIntensity replied:
Done

vstinner commented on Modules/_testcapimodule.c:
```diff
@@ -2546,6 +2546,100 @@ toggle_reftrace_printer(PyObject *ob, PyObject *arg)
+    for (int i = 0; i < 10; ++i) {
+        int res = PyThreadState_Ensure(ref);
+        assert(res == 0);
+        assert(PyInterpreterState_Get() == interp);
```
Can you add the check `PyThreadState_GetUnchecked() != NULL`?

vstinner commented on Modules/_testcapimodule.c:
```diff
@@ -2546,6 +2546,100 @@ toggle_reftrace_printer(PyObject *ob, PyObject *arg)
+
+    for (int i = 0; i < 20; ++i) {
+        PyThreadState_Release();
+    }
```
I suppose that PyThreadState_GetUnchecked() should be NULL at this point? Maybe add a check for that?

ZeroIntensity replied:
I redid this test.

vstinner commented on Programs/_testembed.c:
```diff
@@ -2313,6 +2313,62 @@ test_get_incomplete_frame(void)
     return result;
 }
 
+const char *THREAD_CODE = "import time\n"
```
```suggestion
const char *THREAD_CODE = \
     "import time\n"
     ...
```

vstinner commented on Programs/_testembed.c:
```diff
@@ -2313,6 +2313,62 @@ test_get_incomplete_frame(void)
+        PyInterpreterRef_Close(ref);
+        return -1;
+    }
+    Py_Finalize();
```
It may be safe to add a PyEvent to wait until the thread started before calling Py_Finalize().

ZeroIntensity replied:
We hold a strong interpreter reference, we're implicitly waiting for it to start anyway.

vstinner replied:
Oh you're right, but I always forget the subtle semantics of interpreter strong references. Can you please add a comment explaining that the interpreter reference makes sure that Py_Finalize() doesn't complete until PyInterpreterRef_Close() is called? Or something like that.

ZeroIntensity replied:
Done

vstinner commented on Python/pylifecycle.c:
```diff
@@ -3464,6 +3472,43 @@ wait_for_thread_shutdown(PyThreadState *tstate)
+    PyTime_t wait_ns = 1000 * 1000; // 1 millisecond
+
+    while (true) {
+        if (PyEvent_WaitTimed(&finalizing->finished, wait_ns, 1)) {
```
Would it be possible to implement exponential backoff?

```
PyTime_t wait_max = (PyTime_t)1000 * 1000 * 1000; // 1 second
...
wait_ns *= 2;
wait_ns = Py_MIN(wait_ns, wait_max);
```

I'm not sure what should be maximum. Maybe 100 ms?

vstinner commented on Python/pystate.c:
```diff
@@ -1294,12 +1294,8 @@ interp_look_up_id(_PyRuntimeState *runtime, int64_t requested_id)
-
-   Fail with RuntimeError if the interpreter is not found. */
-
 PyInterpreterState *
```
Can you mark this function as static?

ZeroIntensity replied:
Done

vstinner commented on Python/pystate.c:
```diff
@@ -1726,6 +1733,26 @@ PyThreadState_Clear(PyThreadState *tstate)
+    if (old == 1 && shutting_down_natives(interp)) {
+        _PyEvent_Notify(&finalizing->finished);
+    } else if (old <= 0) {
+        Py_FatalError("interpreter has negative reference count");
```
You may mention PyInterpreterRef_Close().

ZeroIntensity replied:
Done

vstinner commented on Python/pystate.c:
```diff
@@ -3109,3 +3136,231 @@ _Py_GetMainConfig(void)
+{
+    assert(interp != NULL);
+    Py_ssize_t refcount = _Py_atomic_load_ssize_relaxed(&interp->threads.finalizing.countdown);
+    return refcount;
```
```suggestion
    return _Py_atomic_load_ssize_relaxed(&interp->threads.finalizing.countdown);
```

vstinner commented on Python/pystate.c:
```diff
@@ -3109,3 +3136,231 @@ _Py_GetMainConfig(void)
+{
+    PyInterpreterState *interp = (PyInterpreterState *)ref;
+    if (interp == NULL) {
+        Py_FatalError("Got a null interpreter reference, likely due to use after close.");
```
```suggestion
        Py_FatalError("Got a null interpreter reference, likely due to use after PyInterpreterRef_Close()");
```

vstinner commented on Python/pystate.c:
```diff
@@ -3109,3 +3136,231 @@ _Py_GetMainConfig(void)
+}
+
+int
+PyInterpreterRef_Get(PyInterpreterRef *ref_ptr)
```
You may rename the parameter to just "ref".

ZeroIntensity replied:
Done

vstinner commented on Python/pystate.c:
```diff
@@ -3109,3 +3136,231 @@ _Py_GetMainConfig(void)
+PyInterpreterWeakRef_Get(PyInterpreterWeakRef *wref_ptr)
+{
+    PyInterpreterState *interp = PyInterpreterState_Get();
+    _PyInterpreterWeakRef *wref = PyMem_RawMalloc(sizeof(_PyInterpreterWeakRef));
```
You have the GIL, so you can use:

```suggestion
    _PyInterpreterWeakRef *wref = PyMem_Malloc(sizeof(_PyInterpreterWeakRef));
```

(need to update PyInterpreterWeakRef_Close.)

ZeroIntensity replied:
`PyInterpreterWeakRef_Close` can be called without the GIL/attached tstate, so we can't use `PyMem_Free` there.

vstinner replied:
I see. Please add a comment to explain that.

vstinner commented on Python/pystate.c:
```diff
@@ -3109,3 +3136,231 @@ _Py_GetMainConfig(void)
+{
+    _PyInterpreterWeakRef *wref = (_PyInterpreterWeakRef *)wref_handle;
+    if (wref == NULL) {
+        Py_FatalError("Got a null weak interpreter reference, likely due to use after close.");
```
```suggestion
        Py_FatalError("Got a null weak interpreter reference, likely due to use after PyInterpreterWeakRef_Close()");
```

vstinner commented on Python/pystate.c:
```diff
@@ -3109,3 +3136,231 @@ _Py_GetMainConfig(void)
+        /* Interpreter has already finished threads */
+        *strong_ptr = 0;
+        return -1;
+    } else {
```
PEP 7:

```suggestion
    }
    else {
```

ZeroIntensity committed:
Apply suggestions from code review

Co-authored-by: Victor Stinner <vstinner@python.org>

vstinner commented on Programs/_testembed.c:
```diff
@@ -2313,7 +2313,9 @@ test_get_incomplete_frame(void)
-const char *THREAD_CODE = "import time\n"
+const char *THREAD_CODE = \
+     "import time\n"
+     ...
```
I didn't mean to include "..." in the code but indent following lines the same way :-)

ZeroIntensity committed:
Fix failing build.

ZeroIntensity committed:
Rename parameter.

ZeroIntensity committed:
Fix formatting.

ZeroIntensity committed:
Add tstate check.

ZeroIntensity committed:
Move to pycore_pystate.h

ZeroIntensity committed:
Revert "Move to pycore_pystate.h"

This reverts commit 6c4c52bff7371f2a160d3fcc730114269f7e6894.

ZeroIntensity committed:
Use an exponential wait time for the event.

ZeroIntensity committed:
Mark function as static.

ZeroIntensity committed:
Update fatal error message.

ZeroIntensity committed:
Remove incorrect assertion.

ZeroIntensity committed:
Add a comment regarding PyMem_RawMalloc()

vstinner commented on Include/cpython/pystate.h:
```diff
@@ -259,3 +265,45 @@ PyAPI_FUNC(_PyFrameEvalFunction) _PyInterpreterState_GetEvalFrameFunc(
+
+PyAPI_FUNC(int) PyInterpreterRef_Get(PyInterpreterRef *ptr);
+PyAPI_FUNC(PyInterpreterRef) PyInterpreterRef_Dup(PyInterpreterRef ref);
+PyAPI_FUNC(int) PyInterpreterRef_Main(PyInterpreterRef *strong_ptr);
```
```suggestion
PyAPI_FUNC(int) PyInterpreterRef_Get(PyInterpreterRef *ref);
PyAPI_FUNC(PyInterpreterRef) PyInterpreterRef_Dup(PyInterpreterRef ref);
PyAPI_FUNC(int) PyInterpreterRef_Main(PyInterpreterRef *ref);
```

ZeroIntensity committed:
Add a comment.

ZeroIntensity committed:
Update Include/cpython/pystate.h

Co-authored-by: Victor Stinner <vstinner@python.org>

ZeroIntensity committed:
Merge branch 'pep-788-impl' of https://github.com/ZeroIntensity/cpython into pep-788-impl

ZeroIntensity committed:
Move some tests around to prevent exposure of the private API.

ZeroIntensity committed:
Move weakref test to internal C API.

ZeroIntensity committed:
Improve reference counting tests.

ZeroIntensity committed:
Remove dead function.

ZeroIntensity committed:
Add some more tests.

ZeroIntensity committed:
Remove unused variables.

ZeroIntensity committed:
Fix some thread state attachment problems.

ZeroIntensity committed:
Only delete thread states created by PyThreadState_Ensure()

ZeroIntensity committed:
Add a test for crossinterpreter ensures.

ZeroIntensity committed:
Add a test for weak interpreter references.

ZeroIntensity committed:
Fix concurrent shutdown races in PyGILState_Ensure().

ZeroIntensity committed:
Add a test for PyInterpreterRef_Main().

ZeroIntensity committed:
Use PyErr_FormatUnraisable to show signals.

ZeroIntensity committed:
Fix a re-entrancy deadlock.

ZeroIntensity committed:
Remove stupid IDE imports.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Merge branch 'pep-788-impl' of https://github.com/ZeroIntensity/cpython into pep-788-impl

ZeroIntensity committed:
Fix interpreter reference count tests.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Add thread state references to PyThreadState_Ensure() and
PyThreadState_Release()

Per the new updates to PEP-788.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Fix incorrect condition.

ZeroIntensity committed:
Hand off the GIL when waiting on interpreter references.

We should probably use an event instead of a silly loop like this, but I
don't care at the moment.

ZeroIntensity committed:
Use new names from the PEP.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Use the new names for the API.

ZeroIntensity committed:
Update signatures for the new revision.

ZeroIntensity committed:
Finish migration to the new API.

ZeroIntensity committed:
Fix failing tests.

ZeroIntensity committed:
Some touchups.

ZeroIntensity committed:
Fix C analyzer.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Update to use new "PyInterpreterGuard" names.

ZeroIntensity committed:
Fix some other missing cases.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Fix stray changes.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Remove PyThreadView.

ZeroIntensity committed:
Remove opaque pointer types.

ZeroIntensity committed:
Rename WHENCE_GILSTATE to WHENCE_C_API.

ZeroIntensity committed:
Add PyThreadState_EnsureFromView().

ZeroIntensity committed:
Fix compilation errors.

ZeroIntensity committed:
Use a dedicated heap allocation for interpreter guards.

ZeroIntensity committed:
Fix PyThreadState_EnsureFromView() and add a test.

ZeroIntensity committed:
Add a race test for PyThreadState_EnsureFromView.

ZeroIntensity committed:
Use an event for waiting on guards.

ZeroIntensity committed:
Emit a fatal error for CTRL^Cs while waiting on finalization guards.

ZeroIntensity committed:
Merge branch 'main' of https://github.com/python/cpython into pep-788-impl

ZeroIntensity committed:
Close owned guards before deallocating the thread state, not after.
